### ISSUE TITLE
cycle time of vm-kernel-linux-debug-x64-be has regressed by 30%

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#30010]
Compare subsequent builds `#1701` and `#1702`:

https://build.chromium.org/p/client.dart/builders/vm-kernel-linux-debug-x64-be/builds/1701

```
Start	Thu Jun 22 08:34:03 2017
End	Thu Jun 22 11:17:04 2017
Elapsed	2 hrs, 43 mins, 0 secs
```

https://build.chromium.org/p/client.dart/builders/vm-kernel-linux-debug-x64-be/builds/1702

```
Start	Thu Jun 22 13:57:16 2017
End	Thu Jun 22 18:00:27 2017
Elapsed	4 hrs, 3 mins, 10 secs
```

**Comment by mraleph** [comment]
I bisected this to 8b40507be30f54a610b46a2b0b74c5a25868b750

@sivachandra could you please investigate?

/cc @a-siva

**Comment by sivachandra** [comment]
The change which @mraleph points out loads the service isolate also from platform.dill file (before that change, only the main isolate was being loaded from kernel on the dartk bots).

What I find is that `Dart_CreateIsolateFromKernel` takes over 1.2s to create an isolate from platform.dill with a debug build. On the other hand, the same function takes just over 0.15s with a release build. Since the culprit change adds a second call to `Dart_CreateIsolateFromKernel` (this new call and the old call execute one after the other), a regression is expected.  What is perplexing  however is that only the debug bots regressed.

I tried to narrow down further and what I find is that the calls to `ClassFinalizer::ProcessPendingClasses` cumulatively take around 0.7s (60% of time taken by `CreateIsolateFromKernel`) with a debug build, but only around 0.03 seconds (20% of the time taken by `CreateIsolateFromKernel`) with a release build.

@kmillikin, you mentioned that eager finalization of classes could be the culprit. Does my finding above confirm that?

I tried to understand why `ClassFinalizer::ProcessPendingClasses` takes so long on debug builds. I could not come up with anything concrete, but it seems to me that assert blocks such [this](https://github.com/dart-lang/sdk/blob/master/runtime/vm/class_finalizer.cc#L213) can cumulatively contribute to the high run time.
